proof/refine/ARM/Ipc_R.thy

```diff
@@ -4621,94 +4713,164 @@ lemma receiveIPC_corres:
+                                               and (\<lambda>s. \<forall>r\<in>zobj_refs reply_cap. ex_nonz_cap_to r s)
+                                               and (\<lambda>s. sym_refs
+                                                        (\<lambda>p. if p = sender
+                                                             then tcb_non_st_state_refs_of s sender
+                                                             else state_refs_of s p))
+                                               and valid_bound_obj (\<lambda>r s. r \<notin> fst ` replies_with_sc s) replyOpt
+                                               and cte_wp_at (\<lambda>c. c = cap.NullCap) (thread, tcb_cnode_index 3)
+                                               and receive_ipc_preamble_rv reply_cap replyOpt"
+                              in hoare_strengthen_post[rotated])
+                        apply (clarsimp simp: valid_sched_active_sc_valid_refills)
+                        apply (rename_tac opt; case_tac opt; clarsimp simp: obj_at_def is_tcb pred_tcb_at_def)
                        apply (wpsimp wp: do_ipc_transfer_tcb_caps hoare_vcg_ball_lift
                                          valid_bound_obj_lift)
+                      apply (rule_tac Q="\<lambda>ya. (\<lambda>s. tcb_at' sender s \<and>
+                                                         tcb_at' thread s \<and>
+                                                         cur_tcb' s \<and>
+                                                         valid_queues s \<and>
+                                                         valid_queues' s \<and>
+                                                         valid_release_queue s \<and>
+                                                         valid_release_queue' s \<and>
+                                                         valid_objs' s \<and>
+                                                         valid_bound_obj' valid_replies'_sc_asrt replyOpt
+                                                          s \<and>
+                                                         cur_tcb' s \<and>
+                                                         sch_act_wf (ksSchedulerAction s) s)"
+                      in hoare_strengthen_post[rotated])
+                       apply (clarsimp simp: sch_act_wf_weak obj_at'_def projectKOs split: option.split)
                       apply (wpsimp wp: valid_replies'_sc_asrt_lift valid_bound_obj'_lift)
-                      apply (rule_tac Q="\<lambda>_ s. sch_act_wf (ksSchedulerAction s) s"
-                             in hoare_post_imp, erule sch_act_wf_weak)
-                      apply wpsimp
                      apply (wpsimp wp: gts_st_tcb_at)
                     apply wpsimp
-                   apply (wpsimp wp: hoare_vcg_ball_lift valid_bound_obj_lift)
+                   apply_trace (wpsimp wp: hoare_vcg_ball_lift valid_bound_obj_lift)
```

Comment:
nitpick: `apply_trace`

@lsf37 Could we add `apply_trace` to the `diag_words` list of thylint?